framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,417 @@
+package org.checkerframework.common.value;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * The Range Class models an integer interval, such as all integers between 1 and 10, inclusive.
+ * annotation.
+ *
+ * <p>Note that this class is not responsible for detecting incorrect parameters, e.g. the value of
+ * "from" could be greater than the value of "to". This incorrectness would be eventually caught by
+ * {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)} when
+ * creating an annotation from range, which would then be replaced with @UnknownVal.
+ *
+ * @author JasonMrX
+ */
+public class Range {
+
+    /** The value 'from' */
+    public final long from;
+
+    /** The value 'to' */
+    public final long to;
+
+    /**
+     * Constructs a range with its bounds specified by two parameters, "from" and "to".
+     *
+     * <p>Note that it is possible to construct a range with incorrect parameters, e.g. the value of
+     * "from" could be greater than the value to "to". This incorrectness would be caught by {@link
+     * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)}
+     * when creating an annotation from range, which would then be replaced with @UnknownVal.
+     *
+     * @param from the lower bound (inclusive)
+     * @param to the higher bound (inclusive)
+     */
+    public Range(long from, long to) {
+        this.from = from;
+        this.to = to;
+    }
+
+    public Range(long value) {
+        this(value, value);
+    }
+
+    public Range() {
+        this.from = Long.MIN_VALUE;
+        this.to = Long.MAX_VALUE;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[%s..%s]", from, to);
+    }
+
+    /** Returns true if the element is contained in this range. */
+    public boolean contains(int element) {
+        return from <= element && element <= to;
+    }
+
+    /**
+     * Returns the smallest range that includes all values contained in either of the two ranges. We
+     * call this the union of two ranges.
+     *
+     * @param right a range to union with this range
+     * @return a range resulting from the union of the specified range and this range
+     */
+    public Range union(Range right) {
+        long resultFrom = Math.min(from, right.from);
+        long resultTo = Math.max(to, right.to);
+        return new Range(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns the smallest range that includes all values contained in both of the two ranges. We
+     * call this the intersection of two ranges.
+     *
+     * <p>Special case:
+     *
+     * <ul>
+     *   <li>If there is no overlap between two ranges, an incorrect range with from greater than to
```

Comment:
I think it's better to return a special value, such as `Range.NOTHING` or `Range.EMPTY`, to represent an empty range.  (Also, an empty range corresponds to bottom, not to top, in the type hierarchy, and returning a value that represents bottom will enable correct handling of that case.)